src/stream/src/common/log_store/mod.rs

```diff
@@ -0,0 +1,270 @@
+}
+
+pub trait LogStoreFactory {
+    type Reader: LogReader;
+    type Writer: LogWriter;
+
+    fn build(self) -> (Self::Reader, Self::Writer);
+}
+
+pub struct BoundedInMemLogStoreWriter {
+    curr_epoch: Option<u64>,
+
+    init_epoch_tx: Option<oneshot::Sender<u64>>,
+    item_tx: Sender<LogStoreReadItem>,
+    truncated_epoch_rx: UnboundedReceiver<u64>,
+}
+
+#[derive(Eq, PartialEq, Debug)]
+enum LogReaderEpochProgress {
+    Consuming(u64),
+    AwaitingTruncate { sealed_epoch: u64, next_epoch: u64 },
+}
+
+const UNINITIALIZED: LogReaderEpochProgress = LogReaderEpochProgress::Consuming(INVALID_EPOCH);
+
+pub struct BoundedInMemLogStoreReader {
+    epoch_progress: LogReaderEpochProgress,
+
+    init_epoch_rx: Option<oneshot::Receiver<u64>>,
+    item_rx: Receiver<LogStoreReadItem>,
+    truncated_epoch_tx: UnboundedSender<u64>,
+}
+
+pub struct BoundedInMemLogStoreFactory {
+    bound: usize,
+}
+
+impl BoundedInMemLogStoreFactory {
+    pub fn new(bound: usize) -> Self {
+        Self { bound }
+    }
+}
+
+impl LogStoreFactory for BoundedInMemLogStoreFactory {
+    type Reader = BoundedInMemLogStoreReader;
+    type Writer = BoundedInMemLogStoreWriter;
+
+    fn build(self) -> (Self::Reader, Self::Writer) {
+        let (init_epoch_tx, init_epoch_rx) = oneshot::channel();
+        let (item_tx, item_rx) = channel(self.bound);
+        let (truncated_epoch_tx, truncated_epoch_rx) = unbounded_channel();
+        let reader = BoundedInMemLogStoreReader {
+            epoch_progress: UNINITIALIZED,
+            init_epoch_rx: Some(init_epoch_rx),
+            item_rx,
+            truncated_epoch_tx,
+        };
+        let writer = BoundedInMemLogStoreWriter {
+            curr_epoch: None,
+            init_epoch_tx: Some(init_epoch_tx),
+            item_tx,
+            truncated_epoch_rx,
+        };
+        (reader, writer)
+    }
+}
+
+impl LogReader for BoundedInMemLogStoreReader {
+    type InitFuture<'a> = impl Future<Output = LogStoreResult<u64>> + 'a;
+    type NextItemFuture<'a> = impl Future<Output = LogStoreResult<LogStoreReadItem>> + 'a;
+    type TruncateFuture<'a> = impl Future<Output = LogStoreResult<()>> + 'a;
+
+    fn init(&mut self) -> Self::InitFuture<'_> {
+        async {
+            let init_epoch_rx = self
+                .init_epoch_rx
+                .take()
+                .expect("should not init for twice");
+            // TODO: should return the error
+            let epoch = init_epoch_rx.await.expect("should be able to init");
+            assert_eq!(self.epoch_progress, UNINITIALIZED);
+            self.epoch_progress = LogReaderEpochProgress::Consuming(epoch);
+            Ok(epoch)
+        }
+    }
+
+    fn next_item(&mut self) -> Self::NextItemFuture<'_> {
+        async {
+            match self.item_rx.recv().await {
+                Some(item) => {
+                    if let LogStoreReadItem::Barrier {
+                        next_epoch,
```

Comment:
How about passing `EpochPair` to `LogStoreReadItem::Barrier` so that we can validate whether we unexpectedly skip some epoch?

Reply:
nvm, I realize that the `BoundedInMemLogStoreWriter` already has a check for that.